router/sleeve.go

```diff
@@ -0,0 +1,1117 @@
+
+	case FragTestAck:
+		return fwd.handleFragTestAck()
+
+	case PMTUTestAck:
+		return fwd.handlePMTUTestAck(msg)
+
+	default:
+		log.Println(fwd.logPrefix(),
+			"Ignoring unknown control message:", msg[0])
+		return nil
+	}
+}
+
+func (fwd *sleeveForwarder) confirmed() error {
+	log.Debug(fwd.logPrefix(), "confirmed")
+
+	if fwd.heartbeatInterval != 0 {
+		// already confirmed
+		return nil
+	}
+
+	// heartbeatInterval flags that we want to send heartbeats,
+	// even if we don't do sendHeartbeat() yet due to lacking the
+	// remote address.
+	fwd.heartbeatInterval = time.Duration(FastHeartbeat)
+	if fwd.remoteAddr != nil {
+		if err := fwd.sendHeartbeat(); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout = time.NewTimer(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) sendHeartbeat() error {
+	log.Debug(fwd.logPrefix(), "sendHeartbeat")
+
+	// Prime the timer for the next heartbeat.  We don't use a
+	// ticker because the interval is not constant.
+	fwd.heartbeatTimer = setTimer(fwd.heartbeatTimer, fwd.heartbeatInterval)
+
+	buf := make([]byte, EthernetOverhead+8)
+	binary.BigEndian.PutUint64(buf[EthernetOverhead:], fwd.connUID)
+	return fwd.sendSpecial(fwd.crypto.EncDF, fwd.senderDF, buf)
+}
+
+func (fwd *sleeveForwarder) handleHeartbeat(special specialFrame) error {
+	uid := binary.BigEndian.Uint64(special.frame[EthernetOverhead:])
+	if uid != fwd.connUID {
+		return nil
+	}
+
+	log.Debug(fwd.logPrefix(), "handleHeartbeat")
+
+	if fwd.remoteAddr == nil {
+		fwd.setRemoteAddr(special.sender)
+		if fwd.heartbeatInterval != time.Duration(0) {
+			if err := fwd.sendHeartbeat(); err != nil {
+				return err
+			}
+		}
+	} else if !udpAddrsEqual(fwd.remoteAddr, special.sender) {
+		log.Println(fwd.logPrefix(),
+			"Peer UDP address changed to", special.sender)
+		fwd.setRemoteAddr(special.sender)
+	}
+
+	if !fwd.ackedHeartbeat {
+		fwd.ackedHeartbeat = true
+		if err := fwd.sendControlMsg([]byte{HeartbeatAck}); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout.Reset(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) setRemoteAddr(addr *net.UDPAddr) {
+	// Although we don't need to lock when reading remoteAddr,
+	// because this thread is the only one that modifies
+	// remoteAddr, we do need to lock when writing it, because
+	// memory models.
+	fwd.lock.Lock()
+	fwd.remoteAddr = addr
+	fwd.lock.Unlock()
+}
+
+func (fwd *sleeveForwarder) handleHeartbeatAck() error {
+	// The connection is nowregarded as established
+	fwd.notifyEstablished()
+
+	if fwd.heartbeatInterval != SlowHeartbeat {
+		fwd.heartbeatInterval = SlowHeartbeat
+		if fwd.heartbeatTimer != nil {
+			fwd.heartbeatTimer.Reset(fwd.heartbeatInterval)
+		}
+	}
+
+	fwd.fragTestTicker = time.NewTicker(FragTestInterval)
+	if err := fwd.sendFragTest(); err != nil {
+		return err
+	}
+
+	// Send a large frame down the DF channel in order to prompt
+	// PMTU discovery to start.
+	return fwd.sendSpecial(fwd.crypto.EncDF, fwd.senderDF,
```

Comment:
For the default PMTU to be reduced down from 65535 to something sensible we're relying on receiving an error back from the stack in `processSendError`, however the architecture docs state that the reason for the PMTU discovery mechanism is that such messages may not be reliably received. Are we always guaranteed to get an error here even if ICMP has been filtered out because the packet is bigger than the MTU of the egress interface? Or do we not care about defending against ICMP filtering stupidity on the hosts, only that which may be happening in the intervening network?